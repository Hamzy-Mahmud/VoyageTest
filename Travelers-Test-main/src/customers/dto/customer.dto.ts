import { IsString, Min, Max, IsInt } from 'class-validator';

export class CustomerDto {
  @IsString()
  name: string;

  @IsInt()
  @Min(1000000000)
  @Max(9999999999)
  telephone: number;
}
