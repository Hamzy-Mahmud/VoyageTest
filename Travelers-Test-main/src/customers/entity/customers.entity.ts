import { Transaction } from 'src/transactions/entity/transactions.entity';
import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  CreateDateColumn,
  OneToMany,
} from 'typeorm';

@Entity()
export class Customer {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  name: string;

  @Column()
  telephone: number;

  @CreateDateColumn()
  created_date: Date;

  @OneToMany(() => Transaction, (transaction) => transaction.customer_id)
  transactions?: Transaction[];
}
