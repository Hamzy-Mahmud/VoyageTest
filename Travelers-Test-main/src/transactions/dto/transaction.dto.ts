import { IsInt, IsNumber} from 'class-validator';

export class TransactionDto {
  @IsInt()
  customer_id: number;

  @IsNumber()
  amount: number;
}