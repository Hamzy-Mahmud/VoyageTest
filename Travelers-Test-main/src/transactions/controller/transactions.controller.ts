import { Controller, Post, Body, UseGuards } from '@nestjs/common';
import { JwtAuthGuard } from 'src/auth/jwt-auth.guard';
import { TransactionDto } from '../dto/transaction.dto';
import { Transaction } from '../entity/transactions.entity';
import { TransactionsService } from '../service/transactions.service';

@Controller('transactions')
export class TransactionsController {
  constructor(private transactionsService: TransactionsService) {}

  @UseGuards(JwtAuthGuard)
  @Post()
  async createTransactions(@Body() transaction: TransactionDto): Promise<Transaction> {
    return await this.transactionsService.create(transaction);
  }
}
