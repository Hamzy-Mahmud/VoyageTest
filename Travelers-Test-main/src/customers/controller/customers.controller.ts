import { Controller, Get, Post, Body, Param, UseGuards } from '@nestjs/common';
import { JwtAuthGuard } from 'src/auth/jwt-auth.guard';
import { CustomerDto } from '../dto/customer.dto';
import { Customer } from '../entity/customers.entity';
import { CustomersService } from '../service/customers.service';

@Controller('customers')
export class CustomersController {
  constructor(private customerService: CustomersService) {}

  @UseGuards(JwtAuthGuard)
  @Get()
  async getCustomers(): Promise<Customer[]> {
    return await this.customerService.findAll();
  }

  @UseGuards(JwtAuthGuard)
  @Get('/:id')
  async getCustomer(@Param('id') id: number): Promise<Customer> {
    return await this.customerService.findOne(id);
  }

  @UseGuards(JwtAuthGuard)
  @Get('/:id/transactions')
  async getCustomerTransactions(@Param('id') id: number): Promise<Customer> {
    return await this.customerService.findCustomerTransactions(id);
  }

  @UseGuards(JwtAuthGuard)
  @Post()
  async createCustomers(@Body() customer: CustomerDto): Promise<Customer> {
    return await this.customerService.create(customer);
  }
}
