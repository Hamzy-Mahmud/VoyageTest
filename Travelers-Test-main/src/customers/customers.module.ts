import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import { CustomersController } from './controller/customers.controller';
import { Customer } from './entity/customers.entity';
import { CustomersService } from './service/customers.service';

@Module({
  imports: [TypeOrmModule.forFeature([Customer])],
  providers: [CustomersService],
  controllers: [CustomersController],
  exports: [CustomersService],
})
export class CustomersModule {}
