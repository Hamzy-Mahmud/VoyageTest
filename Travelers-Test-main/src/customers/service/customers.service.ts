import { Injectable, HttpStatus, HttpException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Transaction } from 'src/transactions/entity/transactions.entity';
import { Repository, DataSource } from 'typeorm';
import { CustomerDto } from '../dto/customer.dto';
import { Customer } from '../entity/customers.entity';

@Injectable()
export class CustomersService {
  constructor(
    @InjectRepository(Customer)
    public  customerRepository: Repository<Customer>,
    private dataSource: DataSource,
  ) {}

  async create(customer: CustomerDto): Promise<Customer> {
    return await this.customerRepository.save(customer);
  }

  async findAll(): Promise<Customer[]> {
    return await this.dataSource
      .getRepository(Customer)
      .createQueryBuilder('customer')
      .orderBy('id', 'DESC')
      .take(10)
      .getMany();
  }

  async findOne(id: number): Promise<Customer> {
    return await this.customerRepository.findOneBy({ id });
  }

  async findCustomerTransactions(id: number): Promise<Customer> {
    try {
      const customerInformation = await this.dataSource
      .createQueryBuilder(Customer, 'customer')
      .leftJoinAndSelect(
        'customer.transactions',
        'transactions',
        'transactions.customer_id = :id',
        { id },
      )
      .where('transactions.customer_id = :id', { id })
      .orderBy('transactions.id', 'DESC')
      .getOne();

    return this.mapFindCustomerTransactionsResponse(customerInformation);
      
    } catch (error) {
      throw new HttpException({
        status: HttpStatus.BAD_REQUEST,
        error: 'customer_id should exists',
      }, HttpStatus.BAD_REQUEST);
    }
  }

  private mapFindCustomerTransactionsResponse(
    customerInformation: Customer,
  ): Customer {
    customerInformation.transactions = customerInformation.transactions.slice(
      0,
      10,
    );

    return customerInformation;
  }
}
