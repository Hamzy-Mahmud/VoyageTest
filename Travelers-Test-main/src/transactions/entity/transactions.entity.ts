import { Customer } from 'src/customers/entity/customers.entity';
import { Entity, Column, PrimaryGeneratedColumn, CreateDateColumn, ManyToOne } from 'typeorm';

@Entity()
export class Transaction {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne(() => Customer, (customer) => customer.id)
  customer_id: number;

  @Column()
  amount: number;

  @CreateDateColumn()
  created_date: Date;

}