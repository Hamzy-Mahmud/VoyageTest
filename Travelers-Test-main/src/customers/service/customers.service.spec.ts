import { Test, TestingModule } from '@nestjs/testing';
import { CustomersService } from './customers.service';
import {
  getRepositoryToken,
  TypeOrmModule,
  getDataSourceToken,
} from '@nestjs/typeorm';
import { Customer } from '../entity/customers.entity';
import { CustomersController } from '../controller/customers.controller';
import { CustomerDto } from '../dto/customer.dto';
import { Repository, DataSource } from 'typeorm';

// @ts-ignore
export const dataSourceMockFactory: () => MockType<DataSource> = jest.fn(
  () => ({
    getRepository: jest.fn(() => ({
      createQueryBuilder: jest.fn(() => ({
        orderBy: jest.fn(() => ({
          take: jest.fn(() => ({
            getMany: jest.fn().mockResolvedValue([
              {
                name: 'test 4',
                telephone: 8977173447,
                id: 9,
                created_date: new Date('2022-09-11T07:20:19.328Z'),
              },
              {
                name: 'test 3',
                telephone: 8977173007,
                id: 8,
                created_date: new Date('2022-09-11T07:20:19.328Z'),
              },
            ]),
          })),
        })),
      })),
    })),
    createQueryBuilder: jest.fn(() => ({
      leftJoinAndSelect: jest.fn(()=>({
        where : jest.fn(() => ({
          orderBy: jest.fn(()=>({
            getOne: jest.fn().mockResolvedValue({
              "id": 1,
              "name": "test 1",
              "telephone": 3636699999,
              "created_date": "2022-09-09T02:51:44.000Z",
              "transactions": [
                  {
                      "id": 13,
                      "amount": -3736699898,
                      "created_date": "2022-09-09T17:43:26.000Z"
                  },
                  {
                      "id": 10,
                      "amount": 3736699898,
                      "created_date": "2022-09-09T16:47:29.000Z"
                  },
                  {
                      "id": 9,
                      "amount": 3736699898,
                      "created_date": "2022-09-09T16:45:07.000Z"
                  },
                  {
                      "id": 8,
                      "amount": 3736699898,
                      "created_date": "2022-09-09T16:42:45.000Z"
                  },
                  {
                      "id": 7,
                      "amount": 3736699898,
                      "created_date": "2022-09-09T16:38:55.000Z"
                  },
                  {
                      "id": 6,
                      "amount": 3736699898,
                      "created_date": "2022-09-09T16:38:55.000Z"
                  },
                  {
                      "id": 5,
                      "amount": 3736699898,
                      "created_date": "2022-09-09T16:38:55.000Z"
                  },
                  {
                      "id": 4,
                      "amount": 3736699898,
                      "created_date": "2022-09-09T16:38:55.000Z"
                  },
                  {
                      "id": 3,
                      "amount": 3736699898,
                      "created_date": "2022-09-09T16:38:55.000Z"
                  },
                  {
                      "id": 2,
                      "amount": 3736699898,
                      "created_date": "2022-09-09T16:38:55.000Z"
                  },
                  {
                      "id": 1,
                      "amount": 3736699898,
                      "created_date": "2022-09-09T16:38:55.000Z"
                  }
              ]
          })
          }))
        }))
      }))
    })),
  }),
);

export type MockType<T> = {
  [P in keyof T]?: jest.Mock<{}>;
};

export const repositoryMockFactory: () => MockType<Repository<any>> = jest.fn(
  () => ({
    findOne: jest.fn((entity) => entity),
    save: jest.fn().mockResolvedValue({
      name: 'test 4',
      telephone: 8977173447,
      id: 9,
      created_date: new Date('2022-09-11T07:20:19.328Z'),
    }),
    findOneBy: jest.fn().mockResolvedValue({
      name: 'test 1',
      telephone: 4958173447,
      id: 1,
      created_date: new Date('2022-09-11T07:20:19.328Z'),
    }),
    // ...
  }),
);

describe('CustomersService', () => {
  let service: CustomersService;
  let repositoryMock: MockType<Repository<Customer>>;
  let dataSourceMock: MockType<DataSource>;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        TypeOrmModule.forFeature([Customer]),
        TypeOrmModule.forRoot({
          type: 'sqlite',
          database: 'travelers',
          entities: [__dirname + '/**/*.entity{.ts,.js}'],
          synchronize: true,
        }),
      ],
      providers: [
        CustomersService,
        {
          provide: getRepositoryToken(Customer),
          useFactory: repositoryMockFactory,
        },
        {
          provide: getDataSourceToken(),
          useFactory: dataSourceMockFactory,
        },
      ],
      controllers: [CustomersController],
      exports: [CustomersService],
    }).compile();

    service = module.get<CustomersService>(CustomersService);
    repositoryMock = module.get(getRepositoryToken(Customer));
    dataSourceMock = module.get(getDataSourceToken());
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should create a customer ', async () => {
    const body: CustomerDto = {
      name: 'test 4',
      telephone: 8977173447,
    };

    const response: any = {
      name: 'test 4',
      telephone: 8977173447,
      id: 9,
      created_date: new Date('2022-09-11T07:20:19.328Z'),
    };

    const createresponse = await service.create(body);
    expect(createresponse).toMatchObject(response);
  });

  it('should find all costumers ', async () => {
    const findAllReponse = await service.findAll();
    expect(findAllReponse).toMatchSnapshot();
  });
  
  it('should find one customer ', async () => {
    const findOneResponse = await service.findOne(1);
    expect(findOneResponse).toMatchSnapshot();
  });

  it('should return the customer transactions ', async () => {
    const findCustomerTransactionsResponse = await service.findCustomerTransactions(1);
    expect(findCustomerTransactionsResponse).toMatchSnapshot();
  });
});
