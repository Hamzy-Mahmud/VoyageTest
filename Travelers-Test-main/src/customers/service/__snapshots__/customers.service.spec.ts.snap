// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomersService should find all costumers  1`] = `
Array [
  Object {
    "created_date": 2022-09-11T07:20:19.328Z,
    "id": 9,
    "name": "test 4",
    "telephone": 8977173447,
  },
  Object {
    "created_date": 2022-09-11T07:20:19.328Z,
    "id": 8,
    "name": "test 3",
    "telephone": 8977173007,
  },
]
`;

exports[`CustomersService should find one customer  1`] = `
Object {
  "created_date": 2022-09-11T07:20:19.328Z,
  "id": 1,
  "name": "test 1",
  "telephone": 4958173447,
}
`;

exports[`CustomersService should return the customer transactions  1`] = `
Object {
  "created_date": "2022-09-09T02:51:44.000Z",
  "id": 1,
  "name": "test 1",
  "telephone": 3636699999,
  "transactions": Array [
    Object {
      "amount": -3736699898,
      "created_date": "2022-09-09T17:43:26.000Z",
      "id": 13,
    },
    Object {
      "amount": 3736699898,
      "created_date": "2022-09-09T16:47:29.000Z",
      "id": 10,
    },
    Object {
      "amount": 3736699898,
      "created_date": "2022-09-09T16:45:07.000Z",
      "id": 9,
    },
    Object {
      "amount": 3736699898,
      "created_date": "2022-09-09T16:42:45.000Z",
      "id": 8,
    },
    Object {
      "amount": 3736699898,
      "created_date": "2022-09-09T16:38:55.000Z",
      "id": 7,
    },
    Object {
      "amount": 3736699898,
      "created_date": "2022-09-09T16:38:55.000Z",
      "id": 6,
    },
    Object {
      "amount": 3736699898,
      "created_date": "2022-09-09T16:38:55.000Z",
      "id": 5,
    },
    Object {
      "amount": 3736699898,
      "created_date": "2022-09-09T16:38:55.000Z",
      "id": 4,
    },
    Object {
      "amount": 3736699898,
      "created_date": "2022-09-09T16:38:55.000Z",
      "id": 3,
    },
    Object {
      "amount": 3736699898,
      "created_date": "2022-09-09T16:38:55.000Z",
      "id": 2,
    },
  ],
}
`;
